### PR TITLE
Stake pool/withdrawal fee

### PR DESCRIPTION
**Pull request by jon-chuang**
This PR introduces a withdrawal fee for `spl-stake-pool`. 

The withdrawal fee prevents nuisance delegators from performing validator arbitrage.

The following has been achieved:
1. Initialize stake pool with withdrawal fee
2. `SetWithdrawalFee` instruction. 
  a. Checks that fee increases do not exceed a factor of 1.5x. 
  b. If fee starts at 0, fee can only increase up to 0%->0.15% in the following epoch.
3. Tests for new functionality
4. cli functionality for initialize with optional withdrawal_fee (defaults to (0, 0) if not provided)
5. Add `set-withdrawal-fee` functionality to the cli

Sample output from cli:
```
spl-stake-pool create-pool --fee-denominator 1000 --fee-numerator 2 --withdrawal-fee-denominator 1000 --withdrawal-fee-numerator 3 --max-validators 1000`
Creating reserve stake 98pgpFtaxTRGTD5YaQPgdggMQ5g8ujFQF79bKser66fz
Creating mint JGhKEDuQS3tDNLKFTENnX4mwFTTYqQpgM8dX7xdRHWQ
Creating pool fee collection account DHVTmhCb1qGSwk2gLqJ1BMHJwKRFJ9Xiy9UJF18NT9HE
Signature: 3aZYistTpaX8dWfS4i4SM6oLVT24RSEXwjx5Y3rgKHC13CWarC1at8W9e7iFu84f8qvtumWbNRidRheiVpCd9XS9
Creating stake pool Vg1uDgymfa3tkDweYiiHzK5jnWZvprGEp5PQUgvjxek
Signature: 4hcDiXcS2jq3bayTKjAxPGgosWBJ9XABRuxSjdsUCxauSc1NsQtgUGZhcMMET4gJZGshkYZ9NU9BSaj8QC1a7VHr
```

```
spl-stake-pool set-withdrawal-fee 9WHSSh7MSNVDm7H1j51RvsYGRe4j4SSKLrdNTkMtuM1n 15 10000
Signature: 2Ly92Ai1mpd2sLLUefH55ddEr6eU6qwVbydNKtMvg6D92uhPYAfQkzD2XsT4qZGTZCwFBLftUd2W1wYJ87JY9S41
```

@billythedummy

### CODE REVIEW

File: stake-pool/program/src/processor.rs [hunk 1]
```diff
@@ -594,6 +617,16 @@ impl Processor {
...
+        // Numerator should be smaller than or equal to denominator (fee <= 1)
+        if withdrawal_fee.numerator > withdrawal_fee.denominator {
+            return Err(StakePoolError::FeeTooHigh.into());
+        }
+        // To keep u64 arithmetic sane, we don't allow denominators (and consequently numerators)
+        // that are too large
+        if withdrawal_fee.denominator > 1_000_000 {
+            msg!("Withdrawal fee denominator too large");
+            return Err(StakePoolError::InvalidFeeDenominator.into());
+        }
```
Comment (joncinque) [inline stake-pool/program/src/processor.rs:629]: What's the reason for this?  If we want to set a max fee denominator of 1 million, we may as well change the whole thing to a `u32`, have it as a `per million` and not use the `Fee` type at all.  For consistency, I'd recommend sticking with the `Fee` type

File: stake-pool/program/src/processor.rs [hunk 2]
```diff
@@ -1890,6 +1930,7 @@ impl Processor {
         let clock = &Clock::from_account_info(clock_info)?;
         let token_program_info = next_account_info(account_info_iter)?;
         let stake_program_info = next_account_info(account_info_iter)?;
+        let manager_fee_info = next_account_info(account_info_iter)?;
```
Comment (joncinque) [inline stake-pool/program/src/processor.rs:1933]: Don't forget to check that this is the right manager fee account! Otherwise, I can just put my own account in twice

File: stake-pool/program/src/processor.rs [hunk 3]
```diff
@@ -2162,16 +2220,88 @@ impl Processor {
...
+            (
+                stake_pool.withdrawal_fee.numerator,
+                stake_pool.withdrawal_fee.denominator,
+            )
+        };
+
+        // This is always safe since numerator <= denominator <= 1_000_000
+        // and |MAX_WITHDRAWAL_FEE_INCREASE| <= 10
+        if old_num * fee.denominator * MAX_WITHDRAWAL_FEE_INCREASE.numerator
+            < fee.numerator * old_denom * MAX_WITHDRAWAL_FEE_INCREASE.denominator
```
Comment (joncinque) [inline stake-pool/program/src/processor.rs:2279]: Would you mind adding a comment for this math?  Even something simple like:
```
new_fee / old_fee <= MAX_WITHDRAWAL_FEE_INCREASE
```

File: stake-pool/program/src/processor.rs [hunk 4]
```diff
@@ -2245,6 +2375,10 @@ impl Processor {
                 msg!("Instruction: SetStaker");
                 Self::process_set_staker(program_id, accounts)
             }
+            StakePoolInstruction::SetWithdrawalFee { fee } => {
+                msg!("Instruction: SetFee");
```
Comment (joncinque) [inline stake-pool/program/src/processor.rs:2379]:
```suggestion
                msg!("Instruction: SetWithdrawalFee");
```

File: stake-pool/program/src/state.rs [hunk 5]
```diff
@@ -128,19 +137,28 @@ impl StakePool {
...
-        if self.fee.denominator == 0 || reward_lamports == 0 {
+        if reward_lamports == 0 {
             return Some(0);
         }
+        // If by chance the reserve account has lamports > 0 even though
+        // total_stake_lamports == 0, we credit it entirely to the manager,
+        // regardless of the fee.
+        if self.total_stake_lamports == 0 || self.pool_token_supply == 0 {
+            return Some(reward_lamports);
+        }
```
Comment (joncinque) [inline stake-pool/program/src/state.rs:158]: Looks like this is from your other PR -- could you remove it?

File: stake-pool/program/src/state.rs [hunk 6]
```diff
@@ -534,6 +552,20 @@ pub struct Fee {
     pub numerator: u64,
 }
 
+impl Fee {
+    /// Applies the Fee's rates to a given amount, `amt`
+    /// returning the amount to be subtracted from it as fees
+    /// (0 if denominator is 0 or amt is 0),
+    /// or None if overflow occurs
+    pub fn apply(&self, amt: u128) -> Option<u128> {
```
Comment (joncinque) [inline stake-pool/program/src/state.rs:560]: I like this!

nit: this should use `u64` as the interface. since the fee is <= 1, and amounts are only ever u64, it will fit.  If not, it'll correctly blow up

File: stake-pool/program/src/state.rs [hunk 7]
```diff
@@ -534,6 +552,20 @@ pub struct Fee {
...
+        if self.denominator == 0 {
+            return Some(0);
+        }
+        amt.checked_mul(self.numerator as u128)?
+            .checked_div(self.denominator as u128)
```
Comment (joncinque) [inline stake-pool/program/src/state.rs:565]: Actually, this brings up a potential bug! If someone sets their fee quite low, there's a chance that someone can withdraw pico amounts to get this calculation to always be 0.  As such, this should return a minimum of 1 if numerator is > 0

File: stake-pool/program/tests/huge_pool.rs [hunk 8]
```diff
@@ -26,13 +26,13 @@ use {
...
         stake_program,
-        state::{AccountType, StakePool, StakeStatus, ValidatorList, ValidatorStakeInfo},
+        state::{AccountType, Fee, StakePool, StakeStatus, ValidatorList, ValidatorStakeInfo},
         MAX_VALIDATORS_TO_UPDATE, MINIMUM_ACTIVE_STAKE,
     },
     spl_token::state::{Account as SplAccount, AccountState as SplAccountState, Mint},
 };
 
-const HUGE_POOL_SIZE: u32 = 4_000;
+const HUGE_POOL_SIZE: u32 = 3950;
```
Comment (joncinque) [inline stake-pool/program/tests/huge_pool.rs:35]: Ah man, what a shame, this makes me sad.  With no withdrawal fee, we can't get to 4k? That seems weird, unless deserialization of the new fields is that heavy.  Which tests fail at 4k?

File: stake-pool/program/tests/set_fee.rs [hunk 9]
```diff
@@ -167,7 +167,7 @@ async fn fail_bad_fee() {
             let program_error = error::StakePoolError::FeeTooHigh as u32;
             assert_eq!(error_index, program_error);
         }
-        _ => panic!("Wrong error occurs while malicious try to set manager"),
+        _ => panic!("Wrong error occurs when setting fee too high"),
```
Comment (joncinque) [inline stake-pool/program/tests/set_fee.rs:170]: Nice catch :smile:

File: stake-pool/program/tests/withdraw.rs [hunk 10]
```diff
@@ -62,9 +62,9 @@ async fn setup() -> (
     .await
     .unwrap();
 
-    let tokens_to_burn = deposit_info.pool_tokens / 4;
+    let tokens_to_withdraw = deposit_info.pool_tokens / 4;
```
Comment (joncinque) [inline stake-pool/program/tests/withdraw.rs:65]: much better, I never liked `tokens_to_burn` for some reason

**Review by joncinque — COMMENTED** [review]
Lots of great stuff here!  Very excited to see that it's moving forward so quickly

File: stake-pool/program/src/processor.rs (hunk 1 above)
Comment (jon-chuang) [reply stake-pool/program/src/processor.rs:629]: It was to simplify the arithmetic so as not to require checked muls. We need to multiply 3 numbers together to check the fee increase limit, and knowing the bounds means we can be safer in terms of error handling.

File: stake-pool/program/src/state.rs (hunk 7 above)
Comment (jon-chuang) [reply stake-pool/program/src/state.rs:565]: We were thinking that such errors were inconsequential as transaction fees would dominate at that point.

File: stake-pool/program/tests/huge_pool.rs (hunk 8 above)
Comment (jon-chuang) [reply stake-pool/program/tests/huge_pool.rs:35]: The test which fails is upgrade for big_pools. Specifically, it is update pool values which fails. I tested that master fails with pool size 4100, so we didn't regress **that** much... 🙂

File: stake-pool/program/src/instruction.rs [hunk 11]
```diff
@@ -942,6 +959,7 @@ pub fn withdraw(
         AccountMeta::new_readonly(sysvar::clock::id(), false),
         AccountMeta::new_readonly(*token_program_id, false),
         AccountMeta::new_readonly(stake_program::id(), false),
+        AccountMeta::new(*manager_fee_account, false),
```
Comment (jon-chuang) [inline stake-pool/program/src/instruction.rs:962]: I guess I'll make it number 8 here as well

File: stake-pool/program/src/processor.rs (hunk 2 above)
Comment (jon-chuang) [reply stake-pool/program/src/processor.rs:1933]: hmm yeah, it also reminds me, we need to include the manager field in the withdraw command in the cli

File: stake-pool/program/tests/withdraw.rs (hunk 10 above)
Comment (jon-chuang) [reply stake-pool/program/tests/withdraw.rs:65]: credit to @billythedummy

File: stake-pool/program/src/processor.rs [hunk 12]
```diff
@@ -2162,16 +2220,88 @@ impl Processor {
...
+        let mut stake_pool = try_from_slice_unchecked::<StakePool>(&stake_pool_info.data.borrow())?;
+        if !stake_pool.is_valid() {
+            return Err(StakePoolError::InvalidState.into());
+        }
+
+        stake_pool.check_manager(manager_info)?;
+
+        if stake_pool.last_update_epoch < clock.epoch {
+            return Err(StakePoolError::StakeListAndPoolOutOfDate.into());
+        }
```
[the remark this answers is not in the document]
Comment (jon-chuang) [reply stake-pool/program/src/processor.rs:2245]: Actually, we kind of want this. Just like for `set_fee`, if the epoch has changed, we want the manager to only be able to `set_withdrawal_fee` for the following epoch, not the current one.

File: stake-pool/program/tests/set_fee.rs (hunk 9 above)
Comment (jon-chuang) [reply stake-pool/program/tests/set_fee.rs:170]: caught a few more :-)

**Comment by jon-chuang** [comment]
Thanks for the quick review @joncinque , let me know if there are any other issues you spot. We'd be working on the liquid staking instruction next.

@billythedummy , would you be able to review as well when you can spare the time? Much appreciated.

File: stake-pool/program/src/processor.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (billythedummy) [reply stake-pool/program/src/processor.rs:2248]: right, agreed with @jon-chuang. I didn't realize before but without this check it's possible for a manager to set the withdrawal fee for the current epoch instead of the next one by calling `set_withdrawal_fee` before anyone has called `update_stake_pool_balance`

[the remark this answers is not in the document]
Comment (billythedummy) [reply stake-pool/program/src/processor.rs:2248]: we're working on a new branch with the `Fee` struct reworked so that we don't need checked arithmetic operations. In brief here's the proposal:

- `numerator` and `denominator` are private `u64` fields
- following restrictions checked on `Fee` initialization: `denominator` >= 1, `numerator` <= `denominator`, `denominator` < some arbitrary precision such that multiplication and division will never overflow `u64`. We're thinking `1_000_000_000`
- impl `PartialOrd` and overload `std::ops::Mul`

so, basically changing `Fee` into a constrained fraction

File: stake-pool/program/src/processor.rs (hunk 1 above)
Comment (billythedummy) [reply stake-pool/program/src/processor.rs:629]: we're working on a new branch with the `Fee` struct reworked so that we don't need checked arithmetic operations. In brief here's the proposal:

- `numerator` and `denominator` are private `u64` fields
- following restrictions checked on `Fee` initialization: `denominator` >= 1, `numerator` <= `denominator`, `denominator` < some arbitrary precision such that multiplication and division will never overflow `u64`. We're thinking `1_000_000_000`
- impl `PartialOrd` and overload `std::ops::Mul`

so, basically changing `Fee` into a constrained fraction

Comment (jon-chuang) [reply stake-pool/program/src/processor.rs:629]: following some discussion with @billythedummy the approach pursued was a bit over the top. 

Decided to just go with u32 fees as I don't think anyone will ever need more than 9 dps of precision.

File: stake-pool/program/tests/huge_pool.rs (hunk 8 above)
Comment (jon-chuang) [reply stake-pool/program/tests/huge_pool.rs]: hehe clever me noticed that `apply_fee` wasn't inlined, we're back at `4_000` now  :muscle:

File: stake-pool/program/src/processor.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (joncinque) [reply stake-pool/program/src/processor.rs:2246]: Ah right, sounds good then!

File: stake-pool/program/src/state.rs (hunk 7 above)
Comment (joncinque) [reply stake-pool/program/src/state.rs]: Makes sense, all good then

File: stake-pool/program/src/processor.rs [hunk 13]
```diff
@@ -2166,16 +2221,91 @@ impl Processor {
...
+            )
+        } else {
+            (
+                stake_pool.withdrawal_fee.numerator,
+                stake_pool.withdrawal_fee.denominator,
+            )
+        };
+
+        // Check that new_fee / old_fee <= MAX_WITHDRAWAL_FEE_INCREASE
+        // No checks are required as all the variables are u32
```
Comment (joncinque) [inline stake-pool/program/src/processor.rs:2275]: This comment is out of date.  Should we say that we allow this to blow up if a ridiculous fee is put in?

File: stake-pool/program/src/state.rs [hunk 14]
```diff
@@ -103,6 +103,15 @@ pub struct StakePool {
 
     /// Preferred withdraw validator vote account pubkey
     pub preferred_withdraw_validator_vote_address: Option<Pubkey>,
+
+    /// Fee assessed on deposits
+    pub deposit_fee: Fee,
```
Comment (joncinque) [inline stake-pool/program/src/state.rs:108]: Could you add the `u8` deposit referral percentage while we're at it?

File: stake-pool/program/tests/huge_pool.rs [hunk 15]
```diff
@@ -26,13 +26,13 @@ use {
...
         stake_program,
-        state::{AccountType, StakePool, StakeStatus, ValidatorList, ValidatorStakeInfo},
+        state::{AccountType, Fee, StakePool, StakeStatus, ValidatorList, ValidatorStakeInfo},
         MAX_VALIDATORS_TO_UPDATE, MINIMUM_ACTIVE_STAKE,
     },
     spl_token::state::{Account as SplAccount, AccountState as SplAccountState, Mint},
 };
 
-const HUGE_POOL_SIZE: u32 = 4_000;
+const HUGE_POOL_SIZE: u32 = 4000;
```
Comment (joncinque) [inline stake-pool/program/tests/huge_pool.rs:35]: nit for consistency
```suggestion
const HUGE_POOL_SIZE: u32 = 4_000;
```

File: stake-pool/program/tests/set_fee.rs [hunk 16]
```diff
@@ -216,6 +216,6 @@ async fn fail_not_updated() {
             let program_error = error::StakePoolError::StakeListAndPoolOutOfDate as u32;
             assert_eq!(error_index, program_error);
         }
-        _ => panic!("Wrong error occurs while malicious try to set manager"),
+        _ => panic!("Wrong error occurs when stake pool out of date"),
```
Comment (joncinque) [inline stake-pool/program/tests/set_fee.rs:219]: double oops, nice catch!

File: stake-pool/program/tests/set_withdrawal_fee.rs [hunk 17]
```diff
@@ -0,0 +1,407 @@
...
+        .err()
+        .unwrap()
+        .unwrap();
+
+    match error {
+        TransactionError::InstructionError(_, InstructionError::Custom(error_index)) => {
+            let program_error = error::StakePoolError::WrongManager as u32;
+            assert_eq!(error_index, program_error);
+        }
+        _ => panic!("Wrong error occurs while malicious try to set manager"),
```
Comment (joncinque) [inline stake-pool/program/tests/set_withdrawal_fee.rs:210]:
```suggestion
        _ => panic!("Wrong error occurs while signing with the wrong manager"),
```

**Review by joncinque — COMMENTED** [review]
Looking really close! Just some minor things, then this should be good to go

**Review by joncinque — APPROVED** [review]
Looks good to me! Let me know when it's ready from your side and we'll integrate

**Comment by jon-chuang** [comment]
Yep LGTM too after final once-over @joncinque . I'll let you to the honours.
